godot-core/src/task/async_runtime.rs

```diff
@@ -0,0 +1,456 @@
+    // Spawning new tasks is only allowed on the main thread for now.
+    // We can not accept Sync + Send futures since all object references (i.e. Gd<T>) are not thread-safe. So a future has to remain on the
+    // same thread it was created on. Godots signals on the other hand can be emitted on any thread, so it can't be guaranteed on which thread
+    // a future will be polled.
+    // By limiting async tasks to the main thread we can redirect all signal callbacks back to the main thread via `call_deferred`.
+    //
+    // Once thread-safe futures are possible the restriction can be lifted.
+    assert!(
+        crate::init::is_main_thread(),
+        "godot_task can only be used on the main thread!"
+    );
+
+    let (task_handle, godot_waker) = ASYNC_RUNTIME.with_runtime_mut(move |rt| {
+        let task_handle = rt.add_task(Box::pin(future));
+        let godot_waker = Arc::new(GodotWaker::new(
+            task_handle.index,
+            task_handle.id,
+            thread::current().id(),
+        ));
+
+        (task_handle, godot_waker)
+    });
+
+    poll_future(godot_waker);
+    task_handle
+}
+
+/// Handle for an active background task.
+///
+/// This handle provides introspection into the current state of the task, as well as providing a way to cancel it.
+///
+/// The associated task will **not** be canceled if this handle is dropped.
+pub struct TaskHandle {
+    index: usize,
+    id: u64,
+    _no_send_sync: PhantomData<*const ()>,
+}
+
+impl TaskHandle {
+    fn new(index: usize, id: u64) -> Self {
+        Self {
+            index,
+            id,
+            _no_send_sync: PhantomData,
+        }
+    }
+
+    /// Cancels the task if it is still pending and does nothing if it is already completed.
+    pub fn cancel(self) {
+        ASYNC_RUNTIME.with_runtime_mut(|rt| {
+            let Some(task) = rt.tasks.get(self.index) else {
+                // getting the task from the runtime might return None if the runtime has already been deinitialized. In this case we just
+                // ignore the cancel request as the entire runtime has already been canceled.
+                return;
+            };
+
+            let alive = match task.value {
+                FutureSlotState::Empty => {
+                    panic!("Future slot is empty when canceling it! This is a bug!")
+                }
+                FutureSlotState::Gone => false,
+                FutureSlotState::Pending(_) => task.id == self.id,
+                FutureSlotState::Polling => panic!("Can not cancel future from inside it!"),
+            };
+
+            if !alive {
+                return;
+            }
+
+            rt.clear_task(self.index);
+        })
+    }
+
+    /// Synchronously checks if the task is still pending or has already completed.
+    pub fn is_pending(&self) -> bool {
+        ASYNC_RUNTIME.with_runtime(|rt| {
+            let slot = rt.tasks.get(self.index).expect("Slot at index must exist!");
```

Comment:
```suggestion
            let slot = rt.tasks.get(self.index).unwrap_or_else(|| panic!("missing future slot at index {}", self.index));
```

Is this btw. something the user can trigger? If yes, should maybe be annotated in `# Panics` section. If it shouldn't happen (assuming no bugs), then it's fine as-is.

Reply:
No, it's a bug if we actually ever hit this.

Reply:
We might then also consider `unreachable!`, although it's not that important. `expect()` also works in such a case, as it would anyway require more debugging...